File: app/src/main/java/me/jfenn/alarmio/data/preference/TimePreferenceData.java
package me.jfenn.alarmio.data.preference;

import java.util.concurrent.TimeUnit;

import me.jfenn.alarmio.data.PreferenceData;
import me.jfenn.alarmio.dialogs.TimeChooserDialog;
import me.jfenn.alarmio.utils.FormatUtils;

public class TimePreferenceData extends CustomPreferenceData {

    private PreferenceData preference;

    public TimePreferenceData(PreferenceData preference, int name) {
        super(name);
        this.preference = preference;
    }

    @Override
    public String getValueName(ViewHolder holder) {
        String str = FormatUtils.formatMillis((long) preference.getValue(holder.getContext()));
        return str.substring(0, str.length() - 3);
    }

    @Override
    public void onClick(final ViewHolder holder) {
        int seconds = (int) TimeUnit.MILLISECONDS.toSeconds((long) preference.getValue(holder.getContext()));
        int minutes = (int) TimeUnit.SECONDS.toMinutes(seconds);
        int hours = (int) TimeUnit.MINUTES.toHours(minutes);
        minutes %= TimeUnit.HOURS.toMinutes(1);
        seconds %= TimeUnit.MINUTES.toSeconds(1);

        TimeChooserDialog dialog = new TimeChooserDialog(holder.getContext());
        dialog.setDefault(hours, minutes, seconds);
        dialog.setListener((hours1, minutes1, seconds1) -> {
            seconds1 += TimeUnit.HOURS.toSeconds(hours1);
            seconds1 += TimeUnit.MINUTES.toSeconds(minutes1);
            preference.setValue(holder.getContext(), TimeUnit.SECONDS.toMillis(seconds1));
            bindViewHolder(holder);
        });
        dialog.show();
    }

}

File: app/src/main/java/me/jfenn/alarmio/data/preference/TimePreferenceData.kt
package me.jfenn.alarmio.data.preference

import me.jfenn.alarmio.data.PreferenceData
import me.jfenn.alarmio.dialogs.TimeChooserDialog
import me.jfenn.alarmio.utils.FormatUtils
import java.util.concurrent.TimeUnit

/**
 * A preference item that holds / displays a time value.
 */
class TimePreferenceData(private val preference: PreferenceData, name: Int) : CustomPreferenceData(name) {

    override fun getValueName(holder: CustomPreferenceData.ViewHolder): String {
        return FormatUtils.formatMillis(preference.getValue<Long>(holder.context)).run {
            substring(0, length - 3)
        }
    }

    override fun onClick(holder: CustomPreferenceData.ViewHolder) {
        val dialog = run {
            var seconds = TimeUnit.MILLISECONDS.toSeconds(preference.getValue<Long>(holder.context)).toInt()
            var minutes = TimeUnit.SECONDS.toMinutes(seconds.toLong()).toInt()
            val hours = TimeUnit.MINUTES.toHours(minutes.toLong()).toInt()
            minutes %= TimeUnit.HOURS.toMinutes(1).toInt()
            seconds %= TimeUnit.MINUTES.toSeconds(1).toInt()

            TimeChooserDialog(holder.context).apply { setDefault(hours, minutes, seconds) }
        }

        dialog.setListener { hours, minutes, seconds ->
            run {
                seconds + TimeUnit.HOURS.toSeconds(hours.toLong()).toInt() + TimeUnit.MINUTES.toSeconds(minutes.toLong()).toInt()
            }.let { totalSeconds ->
                preference.setValue(holder.context, TimeUnit.SECONDS.toMillis(totalSeconds.toLong()))
                bindViewHolder(holder)
            }
        }
        dialog.show()
    }

}
